MapboxNavigation/NavigationMapView.swift

```diff
@@ -899,40 +1035,122 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
         
         return symbol
     }
-    
-    func routeStyleLayer(identifier: String, source: MGLSource) -> MGLStyleLayer {
-        let line = MGLLineStyleLayer(identifier: identifier, source: source)
-        line.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel)
-        line.lineOpacity = NSExpression(forConditional:
-            NSPredicate(format: "isAlternateRoute == true"),
-                                        trueExpression: NSExpression(forConstantValue: 1),
-                                        falseExpression: NSExpression(forConditional: NSPredicate(format: "isCurrentLeg == true"),
-                                                                      trueExpression: NSExpression(forConstantValue: 1),
-                                                                      falseExpression: NSExpression(forConstantValue: 0)))
-        line.lineColor = NSExpression(format: "TERNARY(isAlternateRoute == true, %@, MGL_MATCH(congestion, 'low' , %@, 'moderate', %@, 'heavy', %@, 'severe', %@, %@))", routeAlternateColor, trafficLowColor, trafficModerateColor, trafficHeavyColor, trafficSevereColor, trafficUnknownColor)
-        line.lineJoin = NSExpression(forConstantValue: "round")
-        
-        return line
+
+    func generateTrafficGradientStops(for route: Route) {
+
+        /**
+         The resulting set of key/value stops that will be used
+         for the route's line gradient.
+         */
+        var stops = [[CGFloat:UIColor]]()
+
+        /**
+         We will keep track of this value as we iterate through
+         the various congestion segments.
+         */
+        var distanceTraveled: CLLocationDistance = 0.0
+
+        /**
+         Begin by calculating individual congestion segments associated
+         with a congestion level, represented as `MGLPolylineFeature`s.
+         */
+        guard let congestionSegments = addCongestion(to: route, legIndex: 0) else { return }
+
+        /**
+         To create the stops dictionary that represents the route line expressed
+         as gradients, for every congestion segment we need one pair of dictionary
+         entries to represent the color to be displayed between that range. Depending
+         on the index of the congestion segment, the pair's first or second key
+         will have a buffer value added or subtracted to make room for a gradient
+         transition between congestion segments.
+
+            green       gradient       red
+                       transition
+         |-----------|~~~~~~~~~~~~|----------|
+         0         0.499        0.501       1.0
+         */
+
+        for (index, line) in congestionSegments.enumerated() {
+            line.getCoordinates(line.coordinates, range: NSMakeRange(0, Int(line.pointCount)))
```

Comment:
This line looks like it replaces `line`’s coordinates with `line`’s coordinates. Can we remove it?

Reply:
Line `1074` is required or else the congestion segment stops don't get calculated correctly.